src/broadcast/broadcast.rs

```diff
@@ -284,21 +383,120 @@ impl<N: NodeIdT> Broadcast<N> {
             // Enqueue a broadcast of a Ready message.
             step.extend(self.send_ready(hash)?);
         }
+        // Upon receiving 2f + 1 matching Ready(h) messages, send full
+        // `Echo` message to every node who hasn't sent us a `CanDecode`
+        if self.count_readys(hash) == 2 * self.netinfo.num_faulty() + 1 {
+            step.extend(self.send_echo_remaining(hash)?);
+        }
+
         Ok(step.join(self.compute_output(hash)?))
     }
 
-    /// Sends an `Echo` message and handles it. Does nothing if we are only an observer.
-    fn send_echo(&mut self, p: Proof<Vec<u8>>) -> Result<Step<N>> {
-        self.echo_sent = true;
+    /// Sends `Echo` message to all left nodes and handles it.
+    fn send_echo_left(&mut self, p: Proof<Vec<u8>>) -> Result<Step<N>> {
         if !self.netinfo.is_validator() {
             return Ok(Step::default());
         }
         let echo_msg = Message::Echo(p.clone());
-        let step: Step<_> = Target::All.message(echo_msg).into();
+        let mut step = Step::default();
+        // `N - 2f + g` node ids to the left of our_id (excluding our_id and proposer_id)
```

Comment:
The proposer isn't currently excluded, is it?
(I think it's better that way: That causes additional complications and should probably be a separate PR, if at all.)

Reply:
No its not :sweat_smile: Forgot to change the comment.